gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,180 @@ func assertAssignableTo(n Node, xt, dt Type, autoNative bool) {
 	}
 }
 
+func assertValidConstExpr(store Store, last BlockNode, n *ValueDecl, expr Expr) {
+	if n.Type != nil {
+		nt := evalStaticType(store, last, n.Type)
+		if xnt, ok := nt.(*NativeType); ok {
+			nt = go2GnoBaseType(xnt.Type)
+		}
+
+		if _, ok := baseOf(nt).(PrimitiveType); !ok {
+			panic(fmt.Sprintf("invalid constant type %s", nt.String()))
+		}
+	}
+
+	nt := evalStaticTypeOf(store, last, expr)
+	if xnt, ok := nt.(*NativeType); ok {
+		nt = go2GnoBaseType(xnt.Type)
+	}
+
+	if nt == nil {
+		panic(fmt.Sprintf("%s (variable of type nil) is not constant", expr))
+	}
+
+	if _, ok := baseOf(nt).(PrimitiveType); !ok {
+		panic(fmt.Sprintf("%s (variable of type %s) is not constant", expr, nt))
+	}
+
+	assertValidConstValue(store, last, expr, nil)
+}
+
+func assertValidConstValue(store Store, last BlockNode, currExpr, parentExpr Expr) {
```

Comment:
At the same time, I’m wondering—many of the checks here are for functions like len and cap. So, is this recursive call really necessary? Could most of the cases be handled under callExpr instead? Please consider this.

Apologies for the continuous iterations in reviewing this PR. Thank you for understanding.